### ISSUE TITLE
Implement Marabou robustness adapter

### ISSUE THREAD

**Issue by stanlrt** [CAIIVS/raitap#129]
# Marabou formal-verification adapter (raitap)

## Context

Issue #126 originally asked for `AbcrownAssessor`. Investigation showed abcrown blocked by `torch<2.9` vs raitap's `torch>=2.10`, plus heavy deps (Gurobi, CUDA 12.8). Marabou (`maraboupy`) is the only currently realistic open-source formal-verification adapter:

- PyPI wheels (`maraboupy>=2.0.0,<3.0`), BSD-style license, no torch coupling
- ONNX-native input — aligns with raitap's existing ONNX backends
- Linux + macOS x86-64; Python 3.8–3.12; no Windows wheels
- Practical scale: ACAS Xu MLPs (sec), small MLPs (min). Cannot verify ResNet/ViT — out of scope

Plan ships full integration: assessor, visualiser, ACAS Xu test fixture, reporting + tracking wiring. Outcome: `uv sync --extra formal-marabou` on Linux/py3.12 enables `MarabouAssessor` end-to-end through raitap's pipeline.

## Decisions (user-resolved)

| Concern | Decision |
|---|---|
| Python | Assume raitap supports py 3.11+ (separate issue, out of scope) |
| Test fixture | ACAS Xu net 1-1 only for CI (smallest VNN-COMP fixture) |
| Scope | Full integration: adapter + e2e + reporting + tracking |
| ONNX bridge | `torch.onnx.export` to tmpfile per `assess()`, cached across samples |
| ONNX export scope | Static-shape MLPs only. `model.eval()`, opset=13, batch=1. Failures propagate as `AssessorBackendIncompatibilityError` (fail loud) |
| Backend acceptance | Both torch and ONNX backends. Torch → export to tmpfile (logged at INFO). ONNX → reuse existing path |
| Counter-example shape | Generic — reshape via `sample.shape`. Works for tabular (5-D vec) and image (4D NCHW) |
| Marabou version pin | `>=2.0.0,<3.0` |
| ACAS Xu provenance | Download via `SAMPLE_SOURCES` with pinned SHA-256, license recorded |
| Output bounds | `lower_bounds=None, upper_bounds=None` in v1. Follow-up issue tracks per-logit min/max queries |
| Budget plumbing | Read from `semantics.budget` passed to `verify_sample`. Reject `norm != LINF` with clear error |
| ONNX export logging | `logger.info` once per `assess()` call when exporting from torch backend |
| CI budget | ≤2 min total. Single config (ACAS Xu net 1-1, VERIFIED + FALSIFIED case). MNIST/image use cases moved to follow-up issue |
| E2E determinism | Tiny eps (1e-5) → VERIFIED; huge eps → FALSIFIED. Both deterministic regardless of runner speed. Timeout 120s |

## Scaffolding already present (no work needed)

- `MethodKind.FORMAL_VERIFICATION`, `VerificationOutcome`, `RobustnessVerdict {VERIFIED, FALSIFIED, UNKNOWN, ERROR}` — `src/raitap/robustness/contracts.py`
- `FormalVerificationAssessor` ABC with full `assess()` framework — `src/raitap/robustness/assessors/base_assessor.py:285`
- `RobustnessResult.output_bounds`, `runtime_per_sample`, formal-verif metrics (`verified_rate`, `falsified_rate`, `unknown_rate`, `error_rate`, `mean_runtime`) — `src/raitap/robustness/results.py`
- Reporting auto-branches on `method_kind` ("Robustness certification" heading), auto-renders `metrics.as_dict()` — `src/raitap/reporting/builder.py:208`
- MLflow auto-logs `RobustnessMetrics` fields via `RobustnessResult.log()` — `src/raitap/robustness/results.py:308`
- Pipeline loop is method-kind agnostic — `src/raitap/run/pipeline.py:149`
- Data layer supports tabular CSV/parquet → `(N, F)` float32 — `src/raitap/data/data.py:37`
- Empty placeholder `src/raitap/robustness/visualisers/formal/__init__.py` ready

### Create

| File | Purpose |
|---|---|
| `src/raitap/robustness/assessors/marabou_assessor.py` | `MarabouAssessor(FormalVerificationAssessor)` |
| `src/raitap/robustness/visualisers/formal/verdict_summary.py` | `VerdictSummaryVisualiser` (bar chart + runtime histogram) |
| `src/raitap/configs/robustness/marabou_linf.yaml` | YAML config for L∞ box verification |
| `src/raitap/configs/data/acas_xu.yaml` | Tabular ACAS Xu data config |
| `src/raitap/robustness/assessors/tests/test_marabou_assessor.py` | Unit tests (mocked maraboupy) |
| `src/raitap/robustness/visualisers/tests/test_verdict_summary.py` | Visualiser snapshot test |
| `src/raitap/robustness/tests/test_e2e_marabou_acas_xu.py` | E2E (`@pytest.mark.e2e`) |
| `src/raitap/reporting/tests/test_formal_section.py` | Reporting renders formal verdict table |

### Edit

| File | Change |
|---|---|
| `src/raitap/robustness/semantics.py` | Add `MARABOU_REGISTRY: Mapping[str, AssessorSemanticsHints]` with `"linf-box"` entry (`MethodKind.FORMAL_VERIFICATION`, `ThreatModel.WHITE_BOX`, `Objective.UNTARGETED`, `PerturbationNorm.LINF`, `families={"smt","complete","sound"}`) |
| `src/raitap/robustness/assessors/__init__.py` | Export `MarabouAssessor` |
| `src/raitap/robustness/visualisers/formal/__init__.py` | Export `VerdictSummaryVisualiser` |
| `src/raitap/configs/schema.py` (line ~52, `DataConfig`) | Add `input_metadata: dict[str, Any] \| None = None` |
| `src/raitap/data/samples.py` | Add `acas_xu_n1_1` (ONNX) + `acas_xu_inputs` (CSV) entries to `SAMPLE_SOURCES` (pin SHA-256, license = BSD-3 Bak et al.) |
| `pyproject.toml` | New extra `formal-marabou = ["maraboupy>=2.0.0,<3.0; python_full_version < '3.13' and sys_platform != 'win32'", "onnx>=1.15"]`. Do NOT add to `robustness` umbrella |
| `.github/workflows/ci.yml` | New lane `os=ubuntu-latest, python=3.12, extras=formal-marabou`, runs `pytest -m e2e -k marabou` |

## Adapter design

```python
class MarabouAssessor(FormalVerificationAssessor):
    algorithm_registry: ClassVar[Mapping[str, AssessorSemanticsHints]] = MARABOU_REGISTRY

    def __init__(self, *, algorithm: str = "linf-box", timeout_s: float = 300.0,
                 epsilon: float = 0.05, norm: str = "Linf", **kwargs: Any) -> None:
        self.algorithm = algorithm
        self.timeout_s = float(timeout_s)
        self.epsilon = float(epsilon)
        self._onnx_cache: dict[tuple[int, tuple[int, ...]], Path] = {}
        self._export_logged: bool = False  # info-log once per assess()

    def check_backend_compat(self, backend: object) -> None:
        # Accept both: torch backends export on demand; ONNX backends reuse onnx_path.
        # No rejection — only fail loud if ONNX export raises later.
        ...

    def _resolve_onnx_path(self, model: nn.Module, sample: torch.Tensor,
                           backend: object | None) -> Path:
        # If backend exposes `onnx_path`, reuse it directly.
        # Else cache key = (id(model), tuple(sample.shape[1:])).
        # On miss: model.eval(); torch.onnx.export(model, sample, tmp,
        #   opset_version=13, do_constant_folding=True, dynamic_axes=None).
        # On torch.onnx.export error → raise AssessorBackendIncompatibilityError
        # with message "Marabou requires static-shape MLPs; export failed: <reason>".
        # First export per assess() call: logger.info(
        #   "Exporting torch model to ONNX for Marabou verification (cached for subsequent samples)").

    def verify_sample(self, model, sample, target, *, budget, backend, **kw) -> VerificationOutcome:
        if budget.norm != PerturbationNorm.LINF:
            raise ValueError(f"MarabouAssessor only supports LINF, got {budget.norm}")
        # 1. Resolve ONNX path (cached)
        # 2. network = Marabou.read_onnx(str(path))
        # 3. inputVars = network.inputVars[0]; outputVars = network.outputVars[0]
        # 4. For each input variable: setLowerBound(v, sample.flat[i] - eps),
        #                              setUpperBound(v, sample.flat[i] + eps)
        # 5. Encode "exists j != target s.t. logit_j - logit_target >= 0"
        #    via addDisjunctionConstraint over per-class inequalities
        # 6. options = Marabou.createOptions(timeoutInSeconds=int(self.timeout_s), verbosity=0)
        # 7. exit_code, vals, stats = network.solve(options=options)
        # 8. Map:
        #    "unsat"           → VERIFIED
        #    "sat"             → FALSIFIED + counter_example reconstructed
        #                        via np.array([vals[v] for v in inputVars.flat]).reshape(sample.shape)
        #    "TIMEOUT"/"UNKNOWN" → UNKNOWN
        # 9. lower_bounds=None, upper_bounds=None (deferred — see follow-up issue)
        # 10. runtime_seconds = stats.getTotalTimeInMicro()/1e6 (or wall clock)
```

## YAML configs

`marabou_linf.yaml`:
```yaml
marabou_linf:
  _target_: MarabouAssessor
  algorithm: linf-box
  constructor:
    epsilon: 0.05
    norm: Linf
    timeout_s: 300
  visualisers:
    - _target_: VerdictSummaryVisualiser
```

`acas_xu.yaml`:
```yaml
name: acas_xu
description: "ACAS Xu collision-avoidance MLP (5→5), VNN-COMP fixture"
source: acas_xu_inputs
input_metadata:
  kind: tabular
  feature_names: [rho, theta, psi, v_own, v_int]
labels:
  source: acas_xu_inputs
  column: label
  encoding: index
```

## Visualiser

`VerdictSummaryVisualiser` mirrors `transparency/visualisers/tabular_visualiser.py` pattern (matplotlib non-image). Two subplots: bar chart of verdict counts; histogram of `runtime_per_sample`. Declares `supported_method_kinds = frozenset({MethodKind.FORMAL_VERIFICATION})`.

## Test strategy

**Unit (no maraboupy install)** — `test_marabou_assessor.py`:
- `test_registry_contains_linf_box`
- `test_factory_parses_marabou_yaml`
- `test_verify_sample_unsat_maps_to_verified` — mock `network.solve` → `("unsat", {}, stats)`
- `test_verify_sample_sat_reconstructs_counter_example_5d` — tabular shape `(1, 5)`
- `test_verify_sample_sat_reconstructs_counter_example_image` — image shape `(1, 1, 28, 28)`
- `test_verify_sample_timeout_maps_to_unknown`
- `test_verify_sample_rejects_non_linf_norm`
- `test_onnx_export_cached_across_samples` — assert `torch.onnx.export` called once for 3 samples
- `test_onnx_export_failure_raises_backend_incompatibility` — mock `torch.onnx.export` → raise
- `test_check_backend_compat_accepts_torch_and_onnx`

**Visualiser** — `test_verdict_summary.py`: snapshot Figure dims + axis labels on synthetic `RobustnessResult` (mix of VERIFIED/FALSIFIED/UNKNOWN/ERROR verdicts).

**Reporting** — `test_formal_section.py`: build `RobustnessResult` with `MethodKind.FORMAL_VERIFICATION`, assert PDF section heading "Robustness certification" + verdict-rate rows.

**E2E (CI ≤2 min)** — `test_e2e_marabou_acas_xu.py`, `@pytest.mark.e2e`, `pytest.importorskip("maraboupy")`:
- `test_acas_xu_verified` — net 1-1, single 5-D input, **eps=1e-5**, timeout=120s. Assert `verdict == VERIFIED`. Expected runtime <30s.
- `test_acas_xu_falsified` — net 1-1, single 5-D input, **eps=10.0** (huge), timeout=120s. Assert `verdict == FALSIFIED`, `counter_example.shape == (1, 5)`. Re-run model on counter-example, assert argmax differs from target. Expected runtime <30s.

CI total budget: 2 tests × ~30s + setup overhead ≈ **<2 min**.

## CI matrix

Add lane to `.github/workflows/ci.yml`:
```yaml
- os: ubuntu-latest
  python-version: "3.12"
  extras: "formal-marabou"
  test-marker: "e2e and marabou"
```
Other lanes unchanged. Marabou wheel install step plus `pytest -m "e2e and marabou"` selects only the two ACAS Xu e2e tests.

## Verification steps (local + CI)

1. `uv sync --extra formal-marabou` on Linux/py3.12 succeeds
2. `uv run pytest src/raitap/robustness -k marabou` green on all platforms (unit, mocked)
3. `uv run pytest -m "e2e and marabou"` green on Linux py3.12 lane only
4. `uv run raitap` with `data: acas_xu`, `robustness: marabou_linf` produces PDF report containing verdict bar chart + MLflow run with `verified_rate` metric
5. `uv run pytest src/raitap/robustness -k "torchattacks or foolbox"` still green (no regression)
6. `uv run ruff check && uv run ruff format --check` clean
7. `uv run pyright src/raitap/robustness/assessors/marabou_assessor.py` clean

## Risk register (ambiguities resolved)

| Risk | Severity | Mitigation |
|---|---|---|
| `torch.onnx.export` fails on dynamic shapes / BatchNorm / unsupported ops | Medium | Force `model.eval()`, opset=13, batch=1. **Fail loud** with `AssessorBackendIncompatibilityError`. Documented MLP-only constraint in adapter docstring. |
| ACAS Xu net license / availability | Low | Source from VNN-COMP 2023 repo (BSD-3, Bak et al.). Pin SHA-256 in `SAMPLE_SOURCES`. Vendor `LICENSE` snippet. Cache to `~/.cache/raitap/acas_xu`. |
| Marabou wheel install fails on CI | Low | `pytest.importorskip("maraboupy")`. Lane has `continue-on-error: false` to fail fast on real install regressions; switch to `true` only if upstream wheel breaks. |
| Counter-example reshape mismatch | Low | Generic `np.array(...).reshape(sample.shape)`. Unit-tested for both 5-D tabular and 4D NCHW. |
| Marabou API drift between minor versions | Low | Pin `>=2.0.0,<3.0`. Unit tests mock the API surface; CI catches drift on real install. |
| GH runner too slow for ACAS Xu | Low | Tiny eps + huge eps cases each run <30s on 4-vCPU runner; 120s timeout = 4× margin. |
| First-time ONNX export silently confusing | Low | `logger.info` on first export per `assess()` call. Visible at default log level. |
| raitap py 3.13 vs maraboupy 3.12 cap | Out of scope | Tracked separately. This plan assumes raitap supports py 3.11+. |

## Effort

~3.5 dev days: adapter+semantics 1d · visualiser+configs 0.5d · ACAS Xu fixture+schema 0.5d · tests 1d · CI+docs 0.5d.

## Out of scope (tracked separately)

- abcrown adapter — #126
- PyRAT adapter (closed-source, license-gated)
- Per-logit output-bound extraction — #131
- ResNet/ViT formal verification (Marabou cannot scale)
- MNIST + LeNet use cases for paper — #132
- Raitap py 3.11+ compatibility — #130
- Reporting redesign (not needed; existing `_build_robustness_section` handles formal kind)
